src/main/sd/init.c

```diff
@@ -0,0 +1,310 @@
+
+    mext2_response* response = send_command(command, MEXT2_R1);
+    if(response == NULL || response -> r1 != R1_IN_IDLE_STATE)
+    {
+        return false;
+    }
+
+    wait_8_clock_cycles((uint8_t*) response);
+    return true;
+}
+
+bool check_voltage_range()
+{
+    //set CMD8
+    uint8_t command_argument[] = {0x00, 0x00, 0x01, 0xaa};
+    set_command(command, COMMAND_CHECK_VOLTAGE, command_argument); 
+
+    mext2_response* response = send_command(command, MEXT2_R7);
+     if(response == NULL)
+    {
+        return false;
+    }
+
+    if(response -> r1 & R1_ILLEGAL_COMMAND)
+        sd_version = SD_V1X;
+
+    wait_8_clock_cycles((uint8_t*) response);
+    return true;
+}
+
+bool read_OCR()
+{
+    //set CMD58
+    uint8_t command_argument[] = {0x00, 0x00, 0x00, 0x00};
+    set_command(command, COMMAND_READ_OCR, command_argument); 
+
+    mext2_response* response = send_command(command, MEXT2_R3);
+    if(response == NULL || (response -> r1 & R1_ILLEGAL_COMMAND) == true)
+    {
+        return false;
+    }
+
+    if(check_sd_version)
+    {
+        if(response -> ocr[0] & (uint8_t)0x40)
+            sd_version = SD_V2XHCXC;
+        else
+            sd_version = SD_V2X;
+    }
+
+    wait_8_clock_cycles((uint8_t*) response);
+    return true;
+}
+
+bool prepare_init_process()
+{
+    //set CMD55
+    uint8_t command_argument[] = {0x00, 0x00, 0x00, 0x00};
+    set_command(command, COMMAND_BEFORE_INIT_PROCESS, command_argument); 
+
+    mext2_response* response = send_command(command, MEXT2_R1);
+    if(response == NULL || response -> r1 & R1_ILLEGAL_COMMAND)
+    {
+        return false;
+    }
+    wait_8_clock_cycles((uint8_t*) response);
+    return true;
+}
+
+uint8_t start_init_process()
+{
+    //set ACMD41
+    uint8_t command_argument[] = {0x00, 0x00, 0x00, 0x00};
+    set_command(command, COMMAND_INIT_PROCESS, command_argument); 
+
+    mext2_response* response = send_command(command, MEXT2_R1);
+    if(response == NULL)
+    {
+        return 0;
+    }
+    
+    return response -> r1;
+}
+
+
+bool read_CSD_register()
+{
+    //set CMD9
+    uint8_t command_argument[] = {0x00, 0x00, 0x00, 0x00};
+    set_command(command, COMMAND_READ_CSD, command_argument); 
+
+    mext2_response* response = send_command(command, MEXT2_R1);
+    if(response == NULL)
+    {
+        return false;
+    }
+
+
+    if(response -> r1 != 0)
+    {
+        debug("Error: reading CSD register failed\n");
+    }
+    else
+    {
+        uint8_t buffer[] = {0xff, 0xff, 0xff, 0xff, 0xff};
+        if(wait_for_response(buffer) == false) 
+            return false;
+
+        memset(csd_register, 0xff, 16);
+        spi_read_write(csd_register, 16);
+        buffer[0] = 0xff;
+        buffer[1] = 0xff;
+        spi_read_write(buffer, 2);
+    }
+
+    wait_8_clock_cycles((uint8_t*) response);
+
+    return true;
+}
+
+uint8_t init()
+{
+    mext2_command* command = (mext2_command*)malloc(sizeof(mext2_command));
```

Comment:
This mallocs with command doesn't look very good. We have to remember about free and all. Maybe use normal variable instead? Is there any reason not to?

Reply:
ok, malloc removed